### DISCUSSION TITLE
ESP32 Audio Kit In/Out Parallel

### DISCUSSION THREAD

**Discussion by pierluigizagaria** [pschatzmann/arduino-audio-tools#212] [Q&A] [answered]
Does anyone know if I can web stream in and out at the same time with an audio kit?
So two streams audiokit-web web-audiokit

**Comment by pschatzmann** [comment]
I am confused: What do you mean with web in and out. 
I suggest that you try to explain what exactly you are trying to do..

But please note that the resources are limited and the more you try to do on the same board, the more careful you need to design your solution and you might e.g. need to decrease the sampling rate.

**Reply by pierluigizagaria** [reply]
I would like to take microphone sound and let it be available from a web stream. 
And the same time play audio to a speaker by a web stream.
In my project I would like to mod my intercom and make it smart. 
To do this I was thinking about using this board to input and outpud audio to the intercom using wifi.

**Reply by pierluigizagaria** [reply]
To be more precise, could the board handle [streams-url_mp3-audiokit](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-url_mp3-audiokit/streams-url_mp3-audiokit.ino) and [streams-audiokit-webserver](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-webserver_aac/streams-audiokit-webserver_aac.ino) together?

**Comment by pschatzmann** [comment, marked as answer]
I have never tried, so I can't give you a final answer: Both the WiFi and I2S can handle multiple input and output at the same time.  
However you will hit __limits__ very quickly so __the only way to know is to try it youself__ and __work on the issues__ that you identify.
- I would expect that you run out of RAM 
   - so you might need to replace the Encoder in the weberserver with a WavEncoder (which uses less RAM)
   - you can try to investigate another decoders 
- If you the audio is breaking up you can decrease  the sampling rate and channels in the webserver
- If the audio is still breaking up you can try to put the copier.copy() or the server->doLoop() into a separate task

Finally I would also expect that the microphone will pick up the audio that you are playing, so this might be disturbing as well.

In a nutshell: it will not be easy and final success is not guaranteed because of the limitations of the Microcontroller
- the RAM is limited
- the CPU speed is relatively low
- the network bandwith is relatively low 

ps. I never looked at the exact memory requirements of the different encoders and decoders, so here you could contribute to the project...
